CRM/Mailing/Service/ListUnsubscribe.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/**
+ * Apply a full range of `List-Unsubscribe` header options.
+ *
+ * @service civi.mailing.listUnsubscribe
+ * @link https://datatracker.ietf.org/doc/html/rfc8058
+ */
+class CRM_Mailing_Service_ListUnsubscribe extends \Civi\Core\Service\AutoService implements \Symfony\Component\EventDispatcher\EventSubscriberInterface {
+
+  public static function getMethods(): array {
+    return [
+      'mailto' => ts('Mailto'),
+      'http' => ts('HTTP(S) Web-Form'),
+      'oneclick' => ts('HTTP(S) One-Click'),
+    ];
+  }
+
+  public static function getSubscribedEvents() {
+    return [
+      '&hook_civicrm_alterMailParams' => ['alterMailParams', 1000],
+    ];
+  }
+
+  /**
+   * @see \CRM_Utils_Hook::alterMailParams()
+   */
+  public function alterMailParams(&$params, $context = NULL): void {
+    if (!in_array($context, ['civimail', 'flexmailer'])) {
```

Comment:
How does context work - ie when would it be flexmailer vs civimail ?

Reply:
CiviMail was traditional and flexmailer is newer but context here could also be single mail, also Message Template is a context as well

Reply:
@seamuslee001 what does 'traditional' mean ? Does it mean the method we use to send SMS that is still in the BAO or something else?

Reply:
FlexMailer ([default header](https://github.com/civicrm/civicrm-core/blob/5.69.0/ext/flexmailer/src/Listener/BasicHeaders.php#L41), [alterMailParams](https://github.com/civicrm/civicrm-core/blob/5.69.0/ext/flexmailer/src/Listener/HookAdapter.php#L31)) vs BAO Engine ([default header](https://github.com/civicrm/civicrm-core/blob/5.69.0/CRM/Mailing/BAO/Mailing.php#L1033), [alterMailParams](https://github.com/civicrm/civicrm-core/blob/5.69.0/CRM/Mailing/BAO/Mailing.php#L1226))

Don't get me wrong -- the idea that BAO engine can still be called is annoying. (See comments on description.) But this issue targets 5.69/5.70 (and it would be unsurprising if folks work on further backports), so I think some "scope discipline" is good.  IMHO, it's easier to have that conditional than to pick up the hunt against BAO.

Reply:
@totten so I think we can rule out the `preview` path - but EventForward STILL calls the old way - that's awful - we REALLY need to get away from this set up where we are still supporting the system we implemented flexmailer to replace!!!

Perhaps we could add some code comments so it's clear that 'civimail' is to be removed as it is only called from one place now

Reply:
Good idea. I had a similar comment a few lines later. Moved it up here and expanded it a bit.